### PR TITLE
Examples: nvcc/cuda 7.5

### PR DESCRIPTION
**Pull request by ax3l**
This PR adds support for binding to nvcc/cuda 7.5 and integrates compile test into travis-ci.

### Dependency Upgrade

`nvcc/7.5` is optional.

Useful CMake C++11/14 handling was added in `CMake/3.1.0` via `CMAKE_CXX_STANDARD` ([Dec 2014](http://www.kitware.com/blog/home/post/812)). Since the old "add flags" approach did not go well with nvcc, this dependency is updated (and it is way more convenient).

I am aware that this version of CMake is still not (yet) shipped with all LTS versions of distributions but I think for a C++11 project it is a very useful starting point:

- [Ubuntu 15.10 (Wily Werewolf)](http://packages.ubuntu.com/wily/cmake): `3.2.2`, latest **stable** Ubuntu release (but not [last LTS release](https://en.wikipedia.org/wiki/List_of_Ubuntu_releases#Ubuntu_14.04_LTS_.28Trusty_Tahr.29) which is *very* old (2011) and will be replaced in 4/2016 with [16.04 LTS  Xenial Xerus](http://packages.ubuntu.com/xenial/cmake) shipping `CMake/3.3.2`)
- [Debian 9.0 (Stretch)](https://packages.debian.org/stretch/cmake): `3.4.1`, still in **testing** but not in *unstable* (shipped in next release)
- [OpenSUSE Leap 42.1](https://build.opensuse.org/package/show/openSUSE:Leap:42.1/cmake): `3.3.2`, latest **stable** OpenSuse release

- Pre-compiled, standalone, binary packages from Kitware: [CMake Download](https://cmake.org/download/)

Not helping users but devs: CMake bootstrap from source is only
```bash
./bootstrap --prefix=/usr
make
sudo make install
```

### Travis Updates

The tests in travis are upgraded to use a newer version of CMake via a ppa (or brew).

For cuda testing, we also use `gcc/4.9` to cover more gcc versions (default for the other linux build stays `gcc/4.8`).

Caching is enabled, which reduces setup times tremendously. For a CUDA upgrade in the future, just delete the cache once after log-in into travis-ci.org and adjust the cuda install line to a newer release download link.

### New Example

A new example `example0.{cu,py,ref}` was added.
This example is build when a new enough version of CUDA is found, otherwise it will simply be ignored.

The lower range of the examples was chosen intentionally since you might want to add further ("normal") examples without needing to skip or rename the cuda example, e.g., in counting test scripts and `examples/example.cpp`.

Also, the example calculates and draws a pretty little [Mandelbrot set](https://en.wikipedia.org/wiki/Mandelbrot_set).

### CUDA Testing

Note that one can install and compile against the nvidia-toolkit without having the nvidia driver installed or cuda capable card detected.

The default mode of operation is for the examples:
  - if nvcc >= 7.5 is found, compile the CUDA example
  - if a device is found, run test on it (else, try to run and do not fail on missing)

`-flto` is not working with `nvcc` and results in a
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: [...]/pybind11/example/example.so: undefined symbol: fatbinData
```
so I disabled it when `nvcc` is detected.

If one wants to *force not use* `nvcc` for the examples, [even when CUDA *would* be available](https://cmake.org/cmake/help/v3.1/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html), configure with:
```
cmake -DCMAKE_DISABLE_FIND_PACKAGE_CUDA=TRUE
```

### CODE REVIEW

File: example/example0.cu [hunk 1]
```diff
@@ -0,0 +1,162 @@
+/*
+    example/example14.cu -- CUDA acceleratred Mandelbrot example
```
Comment (ax3l) [inline example/example0.cu]: argh, outdated.

**Comment by wjakob** [comment]
Sadly it's not an option to switch to CMake > 2.8, since many LTS linux distros still ship with that.

BTW: This is at least currently a fairly big patch -- do you think there is a way to strip it down quite a bit more? (why not use cuda to add two arrays or something, a full-blown mandelbrot seems a bit overkill and not really related to this project)

**Comment by ax3l** [comment]
@wjakob Distros: I just updated my description to cover that.

**Comment by ax3l** [comment]
@wjakob yes, I tried to keep it compact.

Actually it only changes  `+110 −17` lines, `+162` is the mandelbrot :D
I can definitely add a smaller cuda example, sure!

**Comment by wjakob** [comment]
ok -- another comment:

I'm hesitant to add a lot of CUDA code to the main CMakeLists.txt file, which should really serve as a reference to deal with the normal CPU case. The main benefit is to have travis coverage.

It would be preferable, I think, to add everything CUDA related to a 'examples/cuda' directory, including a CMake file with the cuda-related cmake incantations.

**Comment by ax3l** [comment]
separate: I had this implemented in https://github.com/ax3l/pybind11/commit/812f7dde29a77240e5fb80c6d3181b087143fcb0 as a first hack but I somehow liked the idea to have it integrated seamlessly with the other examples and only added as an additional, optional compiler.

If you prefer that I can make it a bash script, remove it from the testing and a two filer, but that is so non generic ;)

**Comment by wjakob** [comment]
I know, but please realize that this is a library for python bindings -- I definitely see the use of testing a different kind of compiler. That said, having GPU support is kind of insane for this kind of project :).

**Comment by ax3l** [comment]
I'll make it pretty and more slick :dancers: 

Is CMake 3.1 (see distros above) ok for you?

**Comment by wjakob** [comment]
Nope, 2.8 please -- I know it's ancient. But consider that the current Ubuntu LTS (Trusty Tahr) still has it. I'm sure RedHat enterprise also has an ancient version.  In the worst case, you could just call GCC directly from travis.

**Comment by ax3l** [comment]
RedHat seems actually to be a problem with [2005, 2007 and 2010 oldstable releases](https://en.wikipedia.org/wiki/Red_Hat_Enterprise_Linux#Product_life_cycle).

But comparing to those we should be aware that at the time of the currently set CMake dependency `2.8` (=`2.8.0+`), [11/2009](http://www.kitware.com/news/home/browse/252?siteid=9), there was not even a C++11 compiler available (nor a final standard; keep in mind C++11 was standardized in 08/2011 :)).

I will send you a PR updating to CMake/3.1.0 as soon as in 4/2016 the next Ubuntu LTS is published. The even older and still supported Ubuntu LTS `precise` ships [gcc/4.6](http://packages.ubuntu.com/precise/gcc) so we can declare `trusty` deprecated at that point too when superseeded by 3 stable and 1 LTS release ;)

One more reason for an early CMake/3.1.0 is actually that we ignore the `gnu-0x/0y` flags (not used in `CMakeLists.txt`) that a `CMake/3.1` can automatically handle.

For old distributions, CMake is [available](https://cmake.org/download/) as one-click, statically linked, precompiled binary (e.g., for installs into $HOME).

With other words: if we want to support old distributions' compilers with adventurous C++11 pre-final implementations let us better choose at least a recent build system that can handle their flags and also can blacklist compiler versions (see: `CMAKE_CXX_COMPILER_VERSION`) that are known not to work (as soon as someone actually tries them).

>  In the worst case, you could just call GCC directly from travis.

You can directly as soon as you add `george-edison55-precise-backports` together with the *two other ppa's* that are already in there that try to fix travis' incredible old linux images :)

**Comment by ax3l** [comment]
will of course honour your wish, just to make the point: I would not recommend that in a C++11 project :)

**Comment by wjakob** [comment]
Sorry for being a pain with all these wishes to work with ancient versions of various packages. :)

**Comment by ax3l** [comment]
that's no problem, I would ask for the same if I see a PR increasing my dependencies :)

**Pull request closed without merging**